### ISSUE TITLE
ABI serializer can't encode optional structs

### ISSUE THREAD

**Issue by jnordberg** [EOSIO/eos#7450]
Given the following ABI:

```json
{
  "version": "eosio::abi/1.1",
  "structs": [
    {
      "name": "foo",
      "fields": [{
          "name": "bar", "type": "name"
        }, {
          "name": "baz", "type": "baz?"
        }]
    },
    {
      "name": "baz",
      "fields": [{
          "name": "qux", "type": "string"
        }]
    }
  ],
  "actions": [{
      "name": "foo", "type": "foo"
    }]
}
```

```
$ cleos --verbose convert pack_action_data mycontract foo '{"bar":"string","baz":null}'
Error 3015004: The type defined in the ABI is invalid
Error Details:
Unknown type baz?
{"type":"baz?","var":null}
{"type":"foo","var":{"bar":"string","baz":null}}
{"type":"foo","var":{"bar":"string","baz":null}}
Stack Trace:
abi_serializer.cpp:494 _variant_to_binary
abi_serializer.cpp:496 _variant_to_binary
abi_serializer.cpp:496 _variant_to_binary
abi_serializer.cpp:510 _variant_to_binary
```

Decoding data with the same ABI works, tested against 1.7.3 and 1.8.0-rc2

**Comment by ismyhc** [comment]
I can confirm this as well through EOS Lynx for iOS. The error occurs when calling rpc endpoint abi_json_to_bin

**Comment by heifner** [comment]
Fixed by #7477
